Fix/graph single api 3239

mpRegalado opened the pull request:
Closes #3239
**Filtered evolution of agents by manager**
Added code to the Evolution graph so that it only shows agents for the currently selected manager.

**Possible issues**
The `wazuh-monitoring-*` indexes only store info about the API by name. For cluster APIs, this name is the name of the cluster; for single manager APIs, this name is the name of the manager.
Should two clusters or two independent managers have the same name, there wouldn't be enough unique information to identify which API they belong to and both would show on the graph.

**Steps to reproduce**
Visit the Agents section in the wazuh menu

**Screenshots**
![Graph of an API with 1 disconnected agent, it does not show the connected agent from the other active API](https://user-images.githubusercontent.com/80431234/118526721-d2f51880-b740-11eb-9a88-7357b4498d9e.png)

> Graph of an API with 1 disconnected agent, it does not show the connected agent from the other active API

![Graph of the other API with the opposite situation](https://user-images.githubusercontent.com/80431234/118526934-089a0180-b741-11eb-82c9-cdc8a6d4ee71.png)
Graph of the other API with the opposite situation

eze9252 reviewed (approved):
LGTM!

CPAlejandro reviewed (approved):
LGTM!

gabiwassan reviewed (approved):
LGTM!

pablomarga reviewed (approved):
LGTM!

Desvelao commented:
Test :x:

In an environment with:
2 managers nodes (1 master and 1 worker):
- Register an agent to the `worker` node.
The indexed document has the `manager` key as the worker, so the filter by `manager` is not useful to solve the issue if only filer by the name of one manager.
![image](https://user-images.githubusercontent.com/34042064/118841919-0793dc00-b8c9-11eb-88b5-b57f7f287376.png)
![image](https://user-images.githubusercontent.com/34042064/118842158-3e69f200-b8c9-11eb-957b-9feb3b09ecc4.png)

Possible solutions:
- Add to the visualization filters, the name of all manager nodes of the API. Something like:
```
{
 "query": {
   "term": {
     "manager.keyword": ["manager-master-node", "manager-worker-node1", "manager-worker-node2", etc...]
   } 
 }
}
```
- Add to the document the `id` of API. This will be a breaking change with the current data indexed to the document in the monitoring indices

Desvelao reviewed (changes requested):
PR requirements:
:x: Add the change to `CHANGELOG.md`

Test :x:

In an environment with:
2 managers nodes (1 master and 1 worker):
- Register an agent to the `worker` node.
The indexed document has the `manager` key as the worker, so the filter by `manager` is not useful to solve the issue if only filer by the name of one manager.
![image](https://user-images.githubusercontent.com/34042064/118841919-0793dc00-b8c9-11eb-88b5-b57f7f287376.png)
![image](https://user-images.githubusercontent.com/34042064/118842158-3e69f200-b8c9-11eb-957b-9feb3b09ecc4.png)

Possible solutions:
- Add to the visualization filters, the name of all manager nodes of the API. Something like:
```
{
 "query": {
   "term": {
     "manager.keyword": ["manager-master-node", "manager-worker-node1", "manager-worker-node2", etc...]
   } 
 }
}
```
- Add to the document the `id` of API. This will be a breaking change with the current data indexed to the document in the monitoring indices

mpRegalado commented:
I implemented the change suggested by @Desvelao and added a function to retrieve the list of managers from the current API to use it in the filter.
![image](https://user-images.githubusercontent.com/80431234/119129356-18626000-ba37-11eb-91e6-5265332e5387.png)

The issue of multiple managers having the same name still exists

frankeros reviewed (approved):
CR: LGTM!

Desvelao commented on public/controllers/agent/components/agents-preview.js:
```diff
@@ -93,7 +94,12 @@ export const AgentsPreview = compose(
     }, {});
   };
-
+  async getManagers() {
+    const agents = await WzRequest.apiReq('GET', '/agents', {});
+    const managers = agents.data.data.affected_items.map((agent) => agent.manager);
+    const uniqueManagers = managers.reduce((unique,manager) => unique.includes(manager) ? unique : [...unique,manager], []);
+    return uniqueManagers;
+  }
```
This could cause bugs due to:
- The request is limited to the first 500 agents. If we need more agents, we need to paginate. It costs so much to get the managers.
- This could work if we get all agents because could exist an agent registered with other managers that are not returned with the request.

Desvelao commented on public/kibana-integrations/kibana-vis.js:
```diff
@@ -215,7 +215,25 @@ class KibanaVis extends Component {
+        },
+        query: {       
+          bool: {
+            should: this.props.managers.map((manager) => {
+              return {term: {"manager.keyword" : manager}}
+            }),
```
You could use the `terms` query instead of multiple `term`.

mpRegalado commented:
Changed the filter so it search by cluster name. This means the conflict may arise now on clusters with same name instead of managers with same name

Desvelao commented on public/controllers/agent/components/agents-preview.js:
```diff
@@ -51,6 +51,7 @@ export const AgentsPreview = compose(
     super(props);
     this.state = { data: [], loading: false, showAgentsEvolutionVisualization: false, agentTableFilters: [] };
     this.wazuhConfig = new WazuhConfig();
+    this.managers = [];
```
Remove if it is not required.

pablomarga reviewed (approved):
CR: :heavy_check_mark: 
Testing: :heavy_check_mark:

CPAlejandro reviewed (commented):
CR: LGTM!
Testing: LGTM!
